src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -613,3 +613,147 @@ void ve2_hwctx_fini(struct amdxdna_ctx *hwctx)
 	ve2_free_hsa_queue(xdna, &hwctx->priv->hwctx_hsa_queue);
 	kfree(hwctx->priv);
 }
+
+static int ve2_update_handshake_pkt(struct amdxdna_ctx *hwctx, u64 paddr, u8 buf_type,
+				    u32 buf_sz, u32 col, bool attach)
+{
+	struct device *aie_dev = hwctx->priv->aie_dev;
+	struct handshake hs = { 0 };
+	int ret = 0;
+
+	WARN_ON(!aie_dev);
+
+	switch (buf_type) {
+	case AMDXDNA_FW_BUF_DEBUG:
+		if (attach) {
+			hs.dbg_buf.dbg_buf_addr_high = upper_32_bits((u64)paddr);
+			hs.dbg_buf.dbg_buf_addr_low = lower_32_bits((u64)paddr);
+			hs.dbg_buf.size = buf_sz;
+		}
+		ret = aie_partition_write_privileged_mem(aie_dev, CERT_HANDSHAKE_OFF(col) +
+							 offsetof(struct handshake,
+								  dbg_buf.dbg_buf_addr_high),
+							 sizeof(hs.dbg_buf), (void *)&hs.dbg_buf);
+		break;
+	case AMDXDNA_FW_BUF_TRACE:
+		if (attach) {
+			hs.trace.dtrace_addr_high = upper_32_bits((u64)paddr);
+			hs.trace.dtrace_addr_low = lower_32_bits((u64)paddr);
+		}
+		ret = aie_partition_write_privileged_mem(aie_dev, CERT_HANDSHAKE_OFF(col) +
+							 offsetof(struct handshake,
+								  trace.dtrace_addr_high),
+							 sizeof(hs.trace), (void *)&hs.trace);
+		break;
+	case AMDXDNA_FW_BUF_LOG:
+		if (attach) {
+			hs.log_addr_high = upper_32_bits((u64)paddr);
+			hs.log_addr_low = lower_32_bits((u64)paddr);
+			hs.log_buf_size = buf_sz;
+		}
+		ret = aie_partition_write_privileged_mem(aie_dev, CERT_HANDSHAKE_OFF(col) +
+							 offsetof(struct handshake, log_addr_high),
+							 sizeof(hs.log_addr_high) +
+							 sizeof(hs.log_addr_high) +
+							 sizeof(hs.log_buf_size),
+							 (void *)&hs.log_addr_high);
+		break;
+	default:
+		ret = -EOPNOTSUPP;
+		break;
+	}
+
+	return ret;
+}
+
+int ve2_hwctx_config(struct amdxdna_ctx *hwctx, u32 type, u64 mdata_hdl, void *buf, u32 size)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_client *client = hwctx->client;
+	struct amdxdna_gem_obj *abo, *mdata_abo;
+	struct fw_buffer_metadata *mdata;
+	u32 prev_buf_sz;
+	u64 buf_paddr;
+	int ret = 0;
+	u32 buf_sz;
+
+	/* Update fw's handshake shared memory with debug/trace buffer details */
+	switch (type) {
+	case DRM_AMDXDNA_CTX_ASSIGN_DBG_BUF:
+		mdata_abo = amdxdna_gem_get_obj(client, mdata_hdl, AMDXDNA_BO_DEV);
+		if (!mdata_abo || !mdata_abo->mem.kva) {
```

Comment:
why do you check kva. It looks kva never used here?

Reply:
Hi Lizhi, it used in below code

mdata = (struct fw_buffer_metadata *)(mdata_abo->mem.kva);